### ISSUE TITLE
Playback freezes for DASH VOD stream with captions enabled

### ISSUE THREAD

**Issue by vpunjabi** [google/shaka-player#2094]
**Have you read the [FAQ](https://bit.ly/ShakaFAQ) and checked for duplicate open issues?**
Yes

**What version of Shaka Player are you using?**
2.5.4

**Can you reproduce the issue with our latest release version?**
Yes

**Can you reproduce the issue with the latest code from `master`?**
NA

**Are you using the demo app or your own custom app?**
Demo

**If custom app, can you reproduce the issue using our demo app?**
Yes

**What browser and OS are you using?**
Chrome on MacOS

**For embedded devices (smart TVs, etc.), what model and firmware version are you using?**
NA

**What are the manifest and license server URIs?**
Sent separately

**What did you do?**
1. Extract media url from json file. Content is DRM protected (widevine).
2. Open Demo app, go to Custom content and enter use media url from step 1.
3. Play content and enable captions.
4. Seek to around 6 minutes into the clip.
5. Notice that playback gets stuck at 6:23.

**What did you expect to happen?**
1. Playback should continue to play past 6:23 whether captions are enabled or not.
2. Ad at 6:23 would play before switching back to content at 6:24.

**What actually happened?**
1. When captions are disabled stream continues to play even after 6:23.  If captions are enabled stream gets stuck at 6:23 and no new segments are downloaded after segment 93.
2. Ads are not DRM protected but content is.  After 6:23 ad should play but ad is skipped.

**Comment by TheJohnBowers** [comment]
I have a demo stream that can reproduce this here:  https://shaka-player-demo.appspot.com/demo/#audiolang=en-US;textlang=en-US;uilang=en-US;asset=https://content.uplynk.com/299ca57da2d04bdaa28c3c643467b3a1.mpd?prettydash=1&vcodec=avc;license=https://content.uplynk.com/wv;panel=CUSTOM%20CONTENT;build=uncompiled -- It appears to be a problem with CC608 embedded in FMP4 and then playing across period boundaries.  In the above example the second period starts at 22 seconds.  So if you play back and turn captions on, the playback hangs at 22 seconds.  If you leave captions turned off it is fine.  If you turn captions on after 22 seconds, playback works fine until a new period is hit.  The error I see in the console is `Assertion failed: All MediaStates should need the same Period or be performing updates.`

**Comment by TheModMaker** [comment]
I've confirmed the bug with your asset.  We have a "fake" stream for embedded text tracks that we use for the track list and StreamingEngine basically ignores it.  Unfortunately it causes problems with Period transitions since it won't try to switch Periods (since it doesn't have any).  So the text stream is not getting updated but the other streams are waiting for a Period transition.

We are currently flattening Periods, and this will probably help with that.  So we will probably be waiting to see if this error persists after that change.

**Comment by AnilPadi** [comment]
Can you please let us know when is the release date for 2.6 Shaka Player? We are looking for this fix as this is breaking our current playback when the captions are turned on.

**Comment by AnilPadi** [comment]
Any update, please?

**Comment by theodab** [comment]
At the moment the 2.6 milestone is 58% complete. We haven't set any hard and fast due date for it, so all I can say is it probably won't be for a little while.
For this issue in specific, we are currently working on flattening periods, as @TheModMaker mentioned. It's been a month since v2.5.4, and period flattening will be a pretty significant change, so we might make a v2.5.5 when it's out, I'm not sure.

**Comment by joeyparrish** [comment]
@AnilPadi, if we can fix this bug without the period flattening feature in v2.6, we could cherry-pick that fix to v2.5.x.  We are about to release v2.5.5 either today or possibly Monday, so the earliest you might see a fix for this issue would be v2.5.6 after that.

**Comment by joeyparrish** [comment]
This could be a duplicate of #2075.  @AnilPadi, @TheJohnBowers, could you please test the fix suggested by @sath33sh in that issue?

```
diff --git a/lib/media/streaming_engine.js b/lib/media/streaming_engine.js
index 2b8840cf..e6d04b1d 100644
--- a/lib/media/streaming_engine.js
+++ b/lib/media/streaming_engine.js
@@ -1406,6 +1406,14 @@ shaka.media.StreamingEngine.prototype.update_ = function(mediaState) {
   // of SegmentReferences as an indicator to determine Period boundaries
   // because a SegmentIndex can provide SegmentReferences outside its Period.
   mediaState.needPeriodIndex = needPeriodIndex;
+  if (mediaState.type == ContentType.VIDEO) {
+    // Update needPeriodIndex of text stream if it is CEA closed captions
+    const textState = this.mediaStates_.get(ContentType.TEXT);
+    if (textState && textState.stream.mimeType ==
+          shaka.util.MimeUtils.CLOSED_CAPTION_MIMETYPE) {
+      textState.needPeriodIndex = needPeriodIndex;
+    }
+  }
   if (needPeriodIndex != currentPeriodIndex) {
     shaka.log.debug(logPrefix,
                     'need Period ' + needPeriodIndex,
```

**Comment by AnilPadi** [comment]
@joeyparrish. 

Thanks for your response. Basically, we would like to have cherry-picked for 2.5.x that way we can ask the CAF receiver team to have this fix included and then we can consume this fix on Chromecast. Currently, we are waiting for CEA-608 closed captions to work on Live and VOD for about a year. Can you please integrate this once @TheJohnBowers confirms.

**Comment by joeyparrish** [comment]
If we can confirm the fix suggested by @sath33sh, we could maybe integrate it into v2.5.5, but definitely by v2.5.6.  I am coordinating updates with the CAF team.

**Comment by joeyparrish** [comment]
Good news!  This was fixed today and the fix will be in v2.5.5.
